### PR TITLE
HPCC-19629 Add thorlib.clusterGroups() / getClusterGroupNames()

### PR DESCRIPTION
**Pull request by jakesmith**
Allows ECL to retrieve list of all groups associated with the
current logical cluster, as a comma separate list.

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by jakesmith** [comment]
@jakesmith - placeholder for now, pending discussion in JIRA.

**Comment by jakesmith** [comment]
@mckellyln - please review

File: ecl/eclagent/eclagent.cpp [hunk 1]
```diff
@@ -2942,13 +2941,8 @@ char * EclAgent::getGroupName()
...
                     {
                         const char *envGroup = envGroups.item(j);
-                        if (groupName.length())
-                        {
-                            if (!strieq(groupName, envGroup))
-                                throw MakeStringException(-1, "getGroupName(): ambiguous groups %s, %s", groupName.str(), envGroup);
-                        }
-                        else
-                            groupName.append(envGroup);
+                        if (!groupNames.contains(envGroup))
```
Comment (mckellyln) [inline ecl/eclagent/eclagent.cpp]: previously it was a stricmp() between goupName and envGroup.  Is contains case-insensitive ?  Does it matter ?

**Review by mckellyln — APPROVED** [review]

**Comment by mckellyln** [comment]
@jakesmith looks fine

File: rtl/include/eclhelper.hpp [hunk 2]
```diff
@@ -702,6 +702,7 @@ interface ICodeContext : public IResourceContext
     virtual char *getClusterName() = 0; // caller frees return string.
     virtual char *getEnv(const char *name, const char *defaultValue) const = 0;
     virtual char *getGroupName() = 0; // caller frees return string.
+    virtual char *getClusterGroupNames() = 0; // caller frees return string.
```
Comment (richardkchapman) [inline rtl/include/eclhelper.hpp]: This is going to break previously compiled queries

**Comment by richardkchapman** [comment]
@jakesmith Compatibility concerns

File: rtl/include/eclhelper.hpp (hunk 2 above)
Comment (jakesmith) [reply rtl/include/eclhelper.hpp]: hm, what is the normal approach?
If I move the method to the end of the interface, that will keep existing methods at same offsets?
Or do I need to up the interface version somewhere (ACTIVITY_INTERFACE_VERSION?) and/or is that enough?
@richardkchapman

Comment (mckellyln) [reply rtl/include/eclhelper.hpp]: Ah, ok I didn't realize this.  So can this only go into a breaking version ?

**Comment by richardkchapman** [comment]
Moving to the end helps. Increasing the version number means that the incompatibility would be reported rather than just causing crashes. But we would much prefer to NOT require recompilation of queries in a minor release, and we would definitely want to avoid it if at all possible in a point release.

@ghalliday Anything to add?

File: rtl/include/eclhelper.hpp (hunk 2 above)
Comment (ghalliday) [reply rtl/include/eclhelper.hpp]: I think if this is added to the end of the interface it should not break existing queries, and you don't need to increase the version number.

**Review by ghalliday — COMMENTED** [review]
@jakesmith otherwise it looks ok

**Comment by jakesmith** [comment]
@ghalliday - have moved getClusterGroupNames() in ICodeContext, but not in implementations/overrides.
Please check.

**Comment by jakesmith** [comment]
@AttilaVamos - is there a ZAP and/or core details for this error?

**Comment by ghalliday** [comment]
Other PRs are also failing with the same issue.  I wonder why it seems to be more common

**Comment by jakesmith** [comment]
From the logs of this crash:
```
00000110 2019-04-09 06:14:00.400 16984 16984 "WARNING: Index layout does not match published layout for index regress::multi::dg_trans_index_evens"
00000111 2019-04-09 06:14:00.400 16984 16984 "Reading file /mnt/disk1/home/vamosax/smoketest/PR-12366/hpcc/var/lib/HPCCSystems/hpcc-data/thor/regress/multi/dg_flat._1_of_4"
00000039 2019-04-09 06:14:00.400 17823 18170 "================================================"
0000003A 2019-04-09 06:14:00.400 17823 18170 "Program:   10.241.40.11:/mnt/disk1/home/vamosax/smoketest/PR-12366/hpcc/opt/HPCCSystems/bin/eclagent"
0000003B 2019-04-09 06:14:00.400 17823 18170 "Signal:    11 Segmentation fault"
0000003C 2019-04-09 06:14:00.400 17823 18170 "Fault IP:  00007FB5240A55B7"
0000003D 2019-04-09 06:14:00.400 17823 18170 "Accessing: 0000000000000000"
0000003E 2019-04-09 06:14:00.400 17823 18170 "Backtrace:"
0000007D 2019-04-09 06:14:00.401 17429 17429 "setResultInt(a3,-3,64)"
0000003F 2019-04-09 06:14:00.401 17823 18170 "  /mnt/disk1/home/vamosax/smoketest/PR-12366/hpcc/opt/HPCCSystems/lib/libhthor.so(_ZN22KeyedLookupPartHandler9doRequestEP8MatchSet+0x97) [0x7fb5240a55b7]"
00000040 2019-04-09 06:14:00.401 17823 18170 "  /mnt/disk1/home/vamosax/smoketest/PR-12366/hpcc/opt/HPCCSystems/lib/libhthor.so(_ZN17PartHandlerThreadI8MatchSet19ThreadedPartHandlerIS0_EE10threadmainEv+0x5d) [0x7fb5240a1e3d]"
00000041 2019-04-09 06:14:00.401 17823 18170 "  /mnt/disk1/home/vamosax/smoketest/PR-12366/hpcc/opt/HPCCSystems/lib/libjlib.so(_ZN20CPooledThreadWrapper3runEv+0x48) [0x7fb51e616688]"
00000042 2019-04-09 06:14:00.401 17823 18170 "  /mnt/disk1/home/vamosax/smoketest/PR-12366/hpcc/opt/HPCCSystems/lib/libjlib.so(_ZN6Thread5beginEv+0x2c) [0x7fb51e60f3fc]"
00000043 2019-04-09 06:14:00.401 17823 18170 "  /mnt/disk1/home/vamosax/smoketest/PR-12366/hpcc/opt/HPCCSystems/lib/libjlib.so(_ZN6Thread11_threadmainEPv+0x1e) [0x7fb51e61096e]"
00000044 2019-04-09 06:14:00.401 17823 18170 "  /lib64/libpthread.so.0(+0x7e25) [0x7fb51cd35e25]"
00000045 2019-04-09 06:14:00.401 17823 18170 "  /lib64/libc.so.6(clone+0x6d) [0x7fb51ca6334d]"
00000046 2019-04-09 06:14:00.401 17823 18170 "Registers:"
00000047 2019-04-09 06:14:00.401 17823 18170 "EAX:00007FB5240A2B50  EBX:0000000001A6F550  ECX:0000000000000000  EDX:0000000000000001  ESI:0000000000000000  EDI:00007FB52431E550"
00000048 2019-04-09 06:14:00.401 17823 18170 "R8 :0000000000000010  R9 :000000000000000B  R10:00007FB4D8000E20  R11:00007FB51CAEEA90"
00000049 2019-04-09 06:14:00.401 17823 18170 "R12:00007FB4D00149D0  R13:00007FB4D0014A80  R14:00007FB4DFFFF700  R15:00007FB4D0013468"
0000004A 2019-04-09 06:14:00.401 17823 18170 "CS:EIP:0033:00007FB5240A55B7"
0000004B 2019-04-09 06:14:00.401 17823 18170 "   ESP:00007FB4DFFFE460  EBP:00007FB4D0014A80"
```

**Comment by jakesmith** [comment]
From the core:
```
Thread 1 (Thread 0x7fb4dffff700 (LWP 18170)):
#0  0x00007fb5240a55b7 in noteEnd (this=0x7fb4d00149d0) at /mnt/disk1/home/vamosax/smoketest/PR-12366/HPCC-Platform/ecl/hthor/hthorkey.cpp:2910
#1  KeyedLookupPartHandler::doRequest (this=0x1a6f550, ms=0x7fb4d0014a80) at /mnt/disk1/home/vamosax/smoketest/PR-12366/HPCC-Platform/ecl/hthor/hthorkey.cpp:3075
#2  0x00007fb5240a1e3d in PartHandlerThread<MatchSet, ThreadedPartHandler<MatchSet> >::threadmain (this=0x7fb4d0013420) at /mnt/disk1/home/vamosax/smoketest/PR-1
2366/HPCC-Platform/ecl/hthor/hthorkey.cpp:1651
#3  0x00007fb51e616688 in CPooledThreadWrapper::run (this=0x7fb4d0013450) at /mnt/disk1/home/vamosax/smoketest/PR-12366/HPCC-Platform/system/jlib/jthread.cpp:908
#4  0x00007fb51e60f3fc in Thread::begin (this=0x7fb4d0013450) at /mnt/disk1/home/vamosax/smoketest/PR-12366/HPCC-Platform/system/jlib/jthread.cpp:267
#5  0x00007fb51e61096e in Thread::_threadmain (v=0x7fb4d0013450) at /mnt/disk1/home/vamosax/smoketest/PR-12366/HPCC-Platform/system/jlib/jthread.cpp:113
#6  0x00007fb51cd35e25 in start_thread () from /lib64/libpthread.so.0
#7  0x00007fb51ca6334d in clone () from /lib64/libc.so.6
rax            0x7fb5240a2b50   140415970454352
rbx            0x1a6f550        27718992
rcx            0x0      0
rdx            0x1      1
rsi            0x0      0
rdi            0x7fb52431e550   140415973057872
rbp            0x7fb4d0014a80   0x7fb4d0014a80
rsp            0x7fb4dfffe460   0x7fb4dfffe460
r8             0x10     16
r9             0xb      11
r10            0x7fb4d8000e20   140414694723104
r11            0x7fb51caeea90   140415847033488
r12            0x7fb4d00149d0   140414560586192
r13            0x7fb4d0014a80   140414560586368
r14            0x7fb4dffff700   140414828934912
r15            0x7fb4d0013468   140414560580712
rip            0x7fb5240a55b7   0x7fb5240a55b7 <KeyedLookupPartHandler::doRequest(MatchSet*)+151>
eflags         0x10246  [ PF ZF IF RF ]
cs             0x33     51
ss             0x2b     43
ds             0x0      0
es             0x0      0
fs             0x0      0
gs             0x0      0
```

**Comment by jakesmith** [comment]
Have moved info + comments to HPCC-20038

**Comment by jakesmith** [comment]
Repushed to force test for clean smoketest.

**Comment by jakesmith** [comment]
@ghalliday - have moved getClusterGroupNames() in ICodeContext, but not in implementations/overrides.
Please check.

**Comment by ghalliday** [comment]
@jakesmith looks good.  Please squash.
Is there a chance to add a regression test - even if it just calls the function and checks that the ClusterGroupName is in the full list?

**Comment by jakesmith** [comment]
@ghalliday - actually, I'm going to close this PR for now. I'm not convinced it's right, or the even the pre-existing thorlib.group() is right.
I was waiting for feedback in the JIRA anyway really, so will wait until then.

**Pull request closed without merging**